core-rust/state-manager/src/mempool/simple_mempool.rs

```diff
@@ -196,52 +187,57 @@ impl SimpleMempool {
         let new_order_data = MempoolDataProposalPriorityOrdering(transaction_data.clone());
 
-        let mut remaining_total_transactions_size = self.remaining_total_transactions_size;
-        let mut remaining_transaction_count = self.remaining_transaction_count;
+        let mut total_transaction_size_free_space = self.remaining_total_transactions_size;
+        let mut total_transaction_count_free_space = self.remaining_transaction_count;
         let mut to_be_removed = Vec::new();
         let mut priority_iter = self.proposal_priority_index.iter();
+        // Collect the lowest priority transactions that are required to be evicted in order to add the new one.
         // Note: worst-case scenario is the biggest transaction that will be rejected against a mempool full of smallest
         // possible transactions. This can be mitigated with a dynamic segment tree which can do fast range sum queries,
         // in order to check rejection before actually getting the evicted transactions. With a minimum transaction of
         // 1024 bytes and current max transaction size of 1MB this should not be a problem yet.
-        while remaining_total_transactions_size < target_remaining_total_transactions_size
-            || remaining_transaction_count < target_remaining_transaction_count
+        while total_transaction_size_free_space < transaction_size
+            || total_transaction_count_free_space < 1
         {
-            let order_data = priority_iter.next_back();
-            match order_data {
+            let lowest_priority_transaction = priority_iter.next();
+            match lowest_priority_transaction {
                 None => {
+                    // Even with an empty mempool we are not able to fulfill the request.
                     panic!("Impossible to add new transaction. Mempool max size lower than transaction size!");
                 }
                 Some(order_data) => {
-                    remaining_total_transactions_size += order_data.0.transaction.raw.0.len();
-                    remaining_transaction_count += 1;
+                    total_transaction_size_free_space +=
+                        order_data.0.transaction.raw.0.len() as u64;
+                    total_transaction_count_free_space += 1;
                     to_be_removed.push(order_data.0.clone());
                 }
             }
         }
 
+        // Check the new transaction is better than all to be removed transactions.
         if !to_be_removed.is_empty() {
             let best_to_be_removed = to_be_removed.last().unwrap();
             if new_order_data
-                > MempoolDataProposalPriorityOrdering((*to_be_removed.last().unwrap()).clone())
+                < MempoolDataProposalPriorityOrdering((*to_be_removed.last().unwrap()).clone())
             {
+                // Note: overflow back to 0 is the desired outcome. This means there is no minimum tip guaranteed to add the transaction.
```

Comment:
Thanks for adding this comment, it draws attention to something I missed :).

But just to say - we should never IMO rely on overflow behaviour - as it panics in debug mode. I would much rather be explicit with eg `wrapping_add`.

Also - I disagree with the premise of using `0` for such a value anyway. We could use an `Option<u16>` where it not being present means there is no valid tip (or even a custom enum) - Rust gives us tools to avoid these magic values, so I feel it's good practice to use them.